Update error message for error 6651 (`Data location must be ...`)

jhb10c opened the pull request:
External and public functions are not allowed to have parameters whose data location is storage. The error message has been updated to reflect this fact. 

This was originally requested in  #[11987](https://github.com/ethereum/solidity/issues/11987)

cameel commented:
Your PR does not actually change the the code that generates the message. It only changes the tests. You need to apply the change here:
https://github.com/ethereum/solidity/blob/d10e668f4f89878983fcaa46e7b88462ffa4380c/libsolidity/analysis/DeclarationTypeChecker.cpp#L370-L391

jhb10c commented:
That makes sense. 

I have added in the line below "string(_variable.isExternalCallableParameter() ? " external" : "")": string(_variable.isPublicCallableParameter() ? " public" : "") and updated the tests using isoltest. This time it should work.

jhb10c commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -383,6 +383,7 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 					string(_variable.isReturnParameter() ? "return " : "") +
 					"parameter in" +
 					string(_variable.isExternalCallableParameter() ? " external" : "") +
+                    string(_variable.isPublicCallableParameter() ? " public" : "") +
```
This adds the word public in front function for error message 6651.

cameel replied:
Your line is indented with spaces while the rest of the code is using tabs. This is probably why [`chk_coding_style`](https://circleci.com/gh/ethereum/solidity/860718) job is failing.

tied2token reviewed (dismissed):
It didn't seem to error for me

jhb10c commented:
@tied2token, I used spaces instead of tabs in the c++ code. This has been updated.

cameel commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -383,6 +383,7 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 					string(_variable.isReturnParameter() ? "return " : "") +
 					"parameter in" +
 					string(_variable.isExternalCallableParameter() ? " external" : "") +
+					string(_variable.isPublicCallableParameter() ? " public" : "") +
```
This is correct but I think we should also be doing this for return parameters. I see a lot of messages in tests that still say "in function" and those are either for storage return parameters or for cases where a parameter has no location at all.

Actually, I see no reason not to make it simple and just always include the visibility in the message. You can get visibility like this (not tested but should work):
```c++
string visibility;
if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_variable.scope()))
    visibility = Declaration::visibilityToString(callable->visibility());
```

chriseth replied:
Just a heads up: If you ask for the visibility of a constructor, you will get a failing assertion.

cameel replied:
ok, yeah, this may need some exta ifs.

jhb10c replied:
Gotcha, @cameel. The first part seems straight forward: update the message for storage return parameters and parameters where type is not defined by including the function type. 

An example of the second type would be like "Data location must be "storage", "memory" or "calldata" for return parameter in function, but none was given." . The function type should be mentioned in the message. 

Thanks for the heads up @chriseth. To check that the CallableDeclaration object is not a constructor, it should look something like:

`string visibility;
if (auto const* callable = dynamic_cast<FunctionDeclaration const*>(_variable.scope()))
{if (!*callable.isConstructor())
       visibility = Declaration::visibilityToString(callable->visibility());
.....
}`

Or would it be better to leave the type as a CallableDeclaration and check that if it is a FunctionDeclaration then check if it is a constructor or not?

cameel replied:
Actually, I see that there's already a check for a constructor parameter in the branch above (`if (_variable.isConstructorParameter())`) so I think you do not have to worry about this case. It's already handled.

cameel reviewed (commented):
Looks good already but we could do more. See my comment below.

Also, please squash your commits so that each one represents a single logical change (with no separate commits for PR fixes).

cameel commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -378,12 +378,17 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
+							"parameter in" +
+							string(visibility.empty() ? visibility : " " + visibility) +
+							" function";
+					}
```
```suggestion
				string visibility;
				if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_variable.scope()))
					visibility = Declaration::visibilityToString(callable->visibility());
				errorString +=
					" for " +
					string(_variable.isReturnParameter() ? "return " : "") +
					"parameter in" +
					string(visibility.empty() ? visibility : " " + visibility) +
					" function";
			}
```

cameel commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -378,12 +378,17 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
+							" for " +
+							string(_variable.isReturnParameter() ? "return " : "") +
+							"parameter in" +
+							string(visibility.empty() ? visibility : " " + visibility) +
```
Since we're touching this code, we could also change the string conversion to a more concise one:
```suggestion
							_variable.isReturnParameter() ? "return "s : ""s +
							"parameter in" +
							visibility.empty() ? visibility : " "s + visibility +
```

cameel commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -378,12 +378,17 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
+							string(_variable.isReturnParameter() ? "return " : "") +
+							"parameter in" +
+							string(visibility.empty() ? visibility : " " + visibility) +
+							" function";
```
We could use our `joinHumanReadable()` to make this more concise (needs `#include <libsolutil/StringUtils.h`). This removes the need to worry about including or omitting the spaces between components:
```suggestion
							string parameterKind = _variable.isReturnParameter() ? "return parameter"s : "parameter"s;
							errorString += " " + joinHumanReadable({"for", parameterKind, "in", visibility, "function"}, " ");
```

jhb10c replied:
I tried using joinHumanReadable() but it could not find the function even with the header included. I have opted to for the second approach. 

There are still whitespace issues that I am fat fingering. ( I am missing how to obtain the correct styling completely....)

cameel replied:
Maybe that's because this file does not use the namespace where the function is defined yet. Please try `util::joinHumanReadable()` or add `using namespace solidity::util;` at the top of the file.

cameel replied:
The whitespace issue seems to be that you're indenting with spaces while we use tabs for .cpp files.

I see that the `chk_coding_style` job detects this but does not print any message about what is wrong. We should change that at some point.

jhb10c replied:
got you, let me try again real fast

cameel reviewed (commented):
Overall the code seems correct. Only style needs some tweaking.

I see one more case where this message does not include visibility: arguments function pointers (see `function_type_argument_external.sol`, `function_type_return_external.sol` and `data_location_in_function_type_fail.sol`). We should fix them too but it's also fine to merge what we have here and do it in a separate PR in my opinion.

Please rebase on latest `develop` - it has fixes for the CI jobs that are failing in your PR. And we need to see that this passes all of them.

leonardoalt commented:
Coding style still needs some fixing.

jhb10c committed:
Update Error Message: Gives function type for data location error

jhb10c commented:
I'm a tech boomer. The whitespace on my IDE was set to space and not tab....

chriseth commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -376,12 +376,17 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 				errorString += " for constructor parameter";
 			else if (_variable.isCallableOrCatchParameter())
+			{
+				string visibility;
```
Maybe
```suggestion
				string functionVisibility;
```
?

jhb10c replied:
Gotcha. I will update this and submit later tonight.

chriseth commented:
Closing due to inactivity. If you still want to work on in, feel free to re-open.
